### PR TITLE
treewide: fail when `nixpkgs.config` is set with explicit `pkgs`, remove all `nixpkgs.config` usages in in-tree modules

### PR DESCRIPTION
**Pull request by K900**
## Description of changes

This gun has been pointed at too many feet now. Kill it.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by Gerg-L — APPROVED** [review]
Very nice!

**Review by alois31 — APPROVED** [review]
I tested this on my system, and everything still works. Although my system config doesn't make use of any of the places where this was previously defined, the diff looks good to me as well.

**Comment by misuzu** [comment]
Maybe add a changelog entry?

**Comment by K900** [comment]
I did?

**Comment by K900** [comment]
Oops. Had it on the wrong branch.

**Review by FRidh — APPROVED** [review]

**Review by misuzu — APPROVED** [review]

**Comment by K900** [comment]
Just realized ofborg doesn't automatically request reviews from affected module maintainers, so I've done that manually. Please check up on your modules and tell me what I did wrong here!

File: pkgs/tools/misc/mlocate/default.nix [hunk 1]
```diff
@@ -1,8 +1,6 @@
-{ lib, stdenv, fetchurl, config }:
+{ lib, stdenv, fetchurl, config, dbfile ? config.locate.dbfile or "/var/cache/locatedb" }:
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/mlocate/default.nix:1]: Tbh I don't know why we have this. Other packages also just hardcode paths.

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/tools/misc/mlocate/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/mlocate/default.nix:1]: Maybe I just create a PR dropping this entirely if @peterhoeg does not object

Comment (K900) [reply pkgs/tools/misc/mlocate/default.nix:1]: I guess someone wanted to put their locate database on some external storage?

**Comment by jtojnar** [comment]
:+1: for removing the vim option from GNOME. It was [introduced](https://github.com/NixOS/nixpkgs/commit/e01c15d433cdaacca0c0525a29ce304b3832d16a) to avoid GTK 2 variant but I would hope that is the default these days. If not, we should probably do that – it would be nice to get GTK 2 out of graphical installation ISO.

**Review by roberth — APPROVED** [review]

**Review by khaneliman — APPROVED** [review]

File: pkgs/tools/misc/mlocate/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/tools/misc/mlocate/default.nix:1]: I think this can be accomplished with symlinks which removes some not really required complexity.

**Comment by arcnmx** [comment]
This likely should've started as a warning; it's a rather broken and incomplete way to solve the problem, with no reasonable workarounds available to use while it's broken. Maybe it could assert that `cfg.config == {} || cfg.config == pkgs.config` to enable workarounds in some way?

**Comment by K900** [comment]
Why is it broken, and why is it an incomplete way to solve the problem?

**Comment by arcnmx** [comment]
I mostly just personally disagree with the strict approach when not everyone necessarily has control over whether a possibly out-of-tree module has decided to set an option under config, though I guess you could `config = mkForce {}` to work around that? It just seems fragile, and the way the meaning and interpretation of the option changes contextually based on how `pkgs` is passed is awkward.

The reason it's broken though is because not all uses of `nixpkgs.config` have been removed here:
- https://github.com/NixOS/nixpkgs/blob/58aa6737113478177ad861c2ff20e10bc8ce12af/nixos/modules/hardware/all-firmware.nix#L72
- https://github.com/NixOS/nixpkgs/blob/58aa6737113478177ad861c2ff20e10bc8ce12af/nixos/modules/programs/firefox.nix#L240
- https://github.com/NixOS/nixpkgs/blob/58aa6737113478177ad861c2ff20e10bc8ce12af/nixos/modules/virtualisation/virtualbox-host.nix#L94

Presumably these need to be changed to use `pkgs.config`? I haven't done a thorough check but that might be all that's left outside of tests using it?

**Comment by K900** [comment]
Read accesses to those shouldn't break anything, as the option isn't removed, and the assertion won't be triggered by _reading_ it. However, yes, those all should be changed to `pkgs.config`, as that is more correct.

**Comment by alois31** [comment]
> It just seems fragile, and the way the meaning and interpretation of the option changes contextually based on how `pkgs` is passed is awkward.

This was already the case before, in fact that's the entire rationale for this PR.

**Comment by arcnmx** [comment]
> Read accesses to those shouldn't break anything, as the option isn't removed, and the assertion won't be triggered by reading it. However, yes, those all should be changed to pkgs.config, as that is more correct.

Oh I mean broken as in *those config options* are broken: if `pkgs` is set explicitly and one wants to set any of the options those module read, you're SoL. There's no migration path available here since the assertion prevents you from setting them temporarily in `nixpkgs.config` to be read by modules as a workaround.
It's not more correct, it's just currently impossible to set both `enableAllFirmware` and `pkgs` since you'll hit a `error: attribute 'allowUnfree' missing` error from it trying to check for that config entry. All reads and writes must be removed from nixpkgs.

**Comment by K900** [comment]
That's actually a completely separate bug.

**Comment by K900** [comment]
(and using `pkgs.config.allowUnfree` does fix it)

**Comment by alyssais** [comment]
The locate change doesn't work at all:

```
       error: attribute 'override' missing

       at /home/qyliss/src/nixpkgs/nixos/modules/misc/locate.nix:233:37:

          232|
          233|     environment.systemPackages = [ (cfg.locate.override { dbfile = cfg.output; }) ];
             |                                     ^
          234|
```

(Config is just `services.locate.enable = true;`.)

**Pull request merged**